src/dynatrace/opentelemetry/metrics/export/serializer.py

```diff
@@ -133,44 +133,104 @@ def _get_metric_key(
+        return DynatraceMetricsSerializer.__normalize_metric_key_section(
+            # delete invalid characters for first section start
+            cls.__re_metric_key_first_identifier_section_start.sub("", section)
         )
 
+    @classmethod
+    def __normalize_metric_key_section(cls, section: str) -> str:
+        # delete invalid characters at the start of the section key
+        section = cls.__re_metric_key_identifier_section_start.sub("", section)
+        section = cls.__re_metric_key_invalid_characters.sub("_", section)
+        return section
+
     @staticmethod
-    def _normalize_metric_key_section(section: str) -> str:
-        section = re.sub("^[^a-zA-Z0-9]+", "", section)
-        section = re.sub("[^a-zA-Z0-9_-]+", "_", section)
+    def _normalize_dimension_key(key: str):
+        # separate sections
+        sections = key.split(".")
+        # normalize them and drop empty sections using filter
+        normalized = list(filter(None, map(
+            DynatraceMetricsSerializer.__normalize_dimension_key_section,
+            sections
+        )))
+
+        return ".".join(normalized)
+
+    # dimension keys have to start with a lowercase letter or an underscore.
+    __re_dimension_key_start = re.compile(r"^[^a-z_]+")
+
+    # other valid characters in dimension keys are lowercase letters, numbers,
+    # colons, underscores and hyphens.
+    __re_dimension_key_invalid_chars = re.compile(r"[^a-z0-9_\-:]+")
+
+    @classmethod
+    def __normalize_dimension_key_section(cls, section: str):
+        # convert to lowercase
+        section = section.lower()
+        # delete leading invalid characters
+        section = cls.__re_dimension_key_start.sub("", section)
+        # replace consecutive invalid characters with one underscore:
+        section = cls.__re_dimension_key_invalid_chars.sub("_", section)
+
         return section
 
     @staticmethod
     def _write_dimensions(
         string_buffer: List[str], dimensions: Iterable[Tuple[str, str]]
     ):
         for key, value in dimensions:
-            string_buffer.append(",")
-            string_buffer.append(key)
-            string_buffer.append("=")
-            string_buffer.append(value)
+            dim_key = DynatraceMetricsSerializer._normalize_dimension_key(key)
+            if dim_key:
+                dim_value = (
+                    DynatraceMetricsSerializer.
+                    _normalize_dimension_value(value))
+
+                string_buffer.append(",")
+                string_buffer.append(dim_key)
+                string_buffer.append("=")
+                string_buffer.append(dim_value)
+
+            # else: the dimension is empty, the dimension is dropped.
 
     @staticmethod
     def _write_timestamp(sb: List[str], aggregator: aggregate.Aggregator):
         sb.append(" ")
         # nanos to millis
         sb.append(str(aggregator.last_update_timestamp // 1000000))
+
+    __re_control_characters = re.compile(r"[\n\t\r]")
```

Comment:
There are a lot more characters than just these three that are non-printing control characters like null, bell, backspace and so on. It might be safer to restrict to a set of allowed characters?

Reply:
i have added a reference to the regex package which actually allows for explicit exclusion of control characters. including a set of characters seems rather tedious, since we would have to include all umlauts and the like for all potential languages. Do you have experience with this package? Or do you think we should address this differently for the sake of not including a third-party package?